control/phaseplot.py

```diff
@@ -1,61 +1,934 @@
+    system.
+
+    Parameters
+    ----------
+    sys : NonlinearIOSystem or callable(t, x, ...)
+        I/O system or function used to generate phase plane data.  If a
+        function is given, the remaining arguments are drawn from the
+        `params` keyword.
+    pointdata : list or 2D array
+        List of the form [xmin, xmax, ymin, ymax] describing the
+        boundaries of the phase plot or an array of shape (N, 2)
+        giving points of at which to plot the vector field.
+    timedata : int or list of int
+        Time to simulate each streamline.  If a list is given, a different
+        time can be used for each initial condition in `pointdata`.
+    gridtype : str, optional
+        The type of grid to use for generating initial conditions:
+        'meshgrid' (default) generates a mesh of initial conditions within
+        the specified boundaries, 'boxgrid' generates initial conditions
+        along the edges of the boundary, 'circlegrid' generates a circle of
+        initial conditions around each point in point data.
+    gridspec : list, optional
+        If the gridtype is 'meshgrid' and 'boxgrid', `gridspec` gives the
+        size of the grid in the x and y axes on which to generate points.
+        If gridtype is 'circlegrid', then `gridspec` is a 2-tuple
+        specifying the radius and number of points around each point in the
+        `pointdata` array.
+    params : dict or list, optional
+        Parameters to pass to system. For an I/O system, `params` should be
+        a dict of parameters and values. For a callable, `params` should be
+        dict with key 'args' and value given by a tuple (passed to callable).
+    color : str
+        Plot the streamlines in the given color.
+    ax : Axes
+        Use the given axes for the plot, otherwise use the current axes.
+
+    Returns
+    -------
+    out : list of Line2D objects
+
+    """
+    # Get system parameters
+    params = kwargs.pop('params', None)
+
+    # Create system from callable, if needed
+    sys = _create_system(sys, params)
+
+    # Parse the arrows keyword
+    arrow_pos, arrow_style = _parse_arrow_keywords(kwargs)
+
+    # Determine the points on which to generate the streamlines
+    points, gridspec = _make_points(pointdata, gridspec, gridtype=gridtype)
+    if dir is None:
+        dir = 'both' if gridtype == 'meshgrid' else 'forward'
+
+    # Create axis if needed
+    if ax is None:
+        ax = plt.gca()
+
+    # Set the axis limits
+    xlim, ylim, maxlim = _set_axis_limits(ax, pointdata)
+
+    # Figure out the color to use
+    color = _get_color(kwargs, ax)
+
+    # Make sure all keyword arguments were processed
+    if check_kwargs and kwargs:
+        raise TypeError("unrecognized keywords: ", str(kwargs))
+
+    # Create reverse time system, if needed
+    if dir != 'forward':
+        revsys = NonlinearIOSystem(
+            lambda t, x, u, params: -np.array(sys.updfcn(t, x, u, params)),
```

Comment:
```suggestion
            lambda t, x, u, params: -np.asarray(sys.updfcn(t, x, u, params)),
```
This won't re-create the array if it is already one. 

relatedly, and this is more of a question: I am confused about the use of each of these methods, all of which do roughly the same thing. Is this correct:  `_rhs` fast and low-level, `updfcn` just a reference to the function that was passed in, and `dynamics`: user-friendly function that may not be as fast? Just checking that this is the right one (speed is kind of important in this application)

Reply:
Correct on the various versions.  In general:
* _updfcn() should not be called directly (though OK here, because we are creating a new system).
* _rhs() should be called inside loops (with a call to _update_params() before the loop starts, which sets the parameter values)
* dynamics() is the user-callable function (which calls _update_params and then _rhs)